devito/dse/rewriters.py

```diff
@@ -405,8 +406,37 @@ def _extract_sum_of_products(self, cluster, template, **kwargs):
 
         return cluster.rebuild(processed)
 
+    @dse_pass
+    def _extract_sum_of_products2(self, cluster, template, **kwargs):
+        """
+        Extract sub-expressions in sum-of-product form, and assign them to temporaries.
+        """
+        make = lambda: Scalar(name=template(), dtype=cluster.dtype).indexify()
+        rule = q_sum_of_product
+        costmodel = lambda e: not (q_leaf(e) or q_terminalop(e))
+        processed, _ = xreplace_constrained(cluster.exprs, make, rule, costmodel)
+
+        return cluster.rebuild(processed)
+
 
 class CustomRewriter(AggressiveRewriter):
 
+    passes_mapper = {
+        'extract_sum_of_products2': AggressiveRewriter._extract_sum_of_products2
+    }
+
+    def __init__(self, passes, template=None, profile=True):
+        try:
+            passes = passes.split(',')
+        except AttributeError:
+            # Already in tuple format
+            if not all(i in CustomRewriter.passes_mapper for i in passes):
+                raise DSEException("Unknown passes `%s`" % str(passes))
+        self.passes = passes
+        self.template = template
+        self.profile = profile
+        super(CustomRewriter, self).__init__()
```

Comment:
Wrong. This should forward template and profile up the chain.

Reply:
```
super(CustomRewriter, self).__init__(self.profile, self.template)
```
?

Reply:
not yet